torch/lowrank.py

```diff
@@ -0,0 +1,325 @@
+"""Implement various linear algebra algorithms for low rank matrices.
+"""
+
+import torch
+
+
+def is_sparse(A):
+    """Check if tensor A is a sparse tensor"""
+    if isinstance(A, torch.Tensor):
+        return A.layout == torch.sparse_coo
+    import scipy.sparse
+    return isinstance(A, scipy.sparse.base.spmatrix)
+
+
+def uniform(low=0.0, high=1.0, size=None, dtype=None, device=None):
+    """Returns a tensor filled with random numbers from a uniform
+    distribution on the interval :math:`[low, high)`
+
+    Arguments::
+
+      size (int...): a sequence of integers defining the shape of the
+        output tensor.  Can be a variable number of arguments or a
+        collection like a list or tuple.
+
+      dtype (:class:`torch.dtype`, optional): the desired data type of
+        returned tensor.  Default: if ``None``, uses a global default
+        (see :func:`torch.set_default_tensor_type`).
+
+      device (:class:`torch.device`, optional): the desired device of
+        returned tensor.  Default: if ``None``, uses the current
+        device for the default tensor type (see
+        :func:`torch.set_default_tensor_type`). :attr:`device` will be
+        the CPU for CPU tensor types and the current CUDA device for
+        CUDA tensor types.
+
+    """
+    attrs = dict(dtype=dtype, device=device)
+    if size is None:
+        r = low + (high - low) * torch.rand(1, **attrs)[0]
+    else:
+        r = low + (high - low) * torch.rand(*size, **attrs)
+    if dtype in [torch.complex32, torch.complex64, torch.complex128]:
+        if size is None:
+            i = low + (high - low) * torch.rand(1, **attrs)[0]
+        else:
+            i = low + (high - low) * torch.rand(*size, **attrs)
+        return r + 1j * i
+    return r
+
+
+def get_floating_dtype(A):
+    """Return the floating point dtype of tensor A.
+    """
+    if isinstance(A, torch.Tensor):
+        index = (0, ) * len(A.shape)
+        return (A.__getitem__(index) * 1.0).dtype
+    return A.dtype.type
+
+
+def get_matmul(A):
+    """Return matrix multiplication implementation.
+    """
+    if A is None:  # A is identity
+        return lambda A, other: other
+    if isinstance(A, torch.Tensor):
+        if is_sparse(A):
+            return torch.sparse.mm
+        else:
+            return torch.matmul
+    else:
+        import numpy
```

Comment:
I'm assuming numpy/scipy are imported to have an implementation that works on tensors from pytorch, numpy, and scipy, is that correct?

@ezyang -- What's our stance on importing numpy/scipy.sparse? I don't see many imports in our codebase, so I assume we do not want to do that.

Reply:
Note that when one does not use numpy or scipy.sparse arrays but only pytorch tensors, then neither numpy nor scipy.sparse will be imported.

In fact, when an user uses the given methods with numpy/scipy.sparse arrays, the corresponding modules are already in `sys.modules`.

Reply:
numpy and scipy imports are eliminated.